### PR TITLE
import `wasi-common` and adapt it for use in host implementation

### PR DESCRIPTION
**Pull request by dicej**
Per recent discussion with Dan and Pat, this begins the work of adapting
`wasi-common` for use in the Preview 2 host implementation.  It removes the
`wiggle` dependency and the rights system, and adds a few structs and fields
specific to Preview 2.  Further changes will follow as we implement more
features.

The intention is that this version of `wasi-common` will be merged back into
`wasmtime` and be used there for both the Preview 1 and Preview 2 WASI host
implementations.  In addition, we would like it to remain runtime-agnostic
(i.e. not depend on `wasmtime` directly or indirectly) so it can be reused by
other runtimes, which was the original intention behind `wasi-common`.
Currently, this requires a manually-written conversion from `wasi_common::Errno`
to `wit_bindgen_host_wasmtime_rust::Errno`; there might be a way to automate
that conversion as part of code generation later on.

This also enables `async` code generation, since most of the `wasi-common`
methods are `async` (though none of them currently `await` anywhere, and many
block).  The intention is that only `wasi-poll::poll-oneoff` will actually
`await`, and we'll replace the current schedulers in `wasi-common` with
something based on e.g. `tokio`.

Other minor changes:

- `wasi-clocks::wall-clock-resolution` now returns an `instant` instead of a
  `datetime` (I presume the latter was a mistake)
- The structs in `host::clocks` have been moved to `wasi-common::clocks`.
- `host::WasiCtx` has been replaced by `wasi-common::WasiCtx`.

Finally, this implements `wasi-filesystem::pread` and `wasi-filesystem::pwrite`
using `wasi-common::WasiFile`.  Implementing the rest of the interface should be
straightforward.

Note that I've deliberately split this into two commits to make reviewing the
`wasi-common` changes easier.  We can squash them into one when merging.

### CODE REVIEW

**Comment by dicej** [comment]
@sunfishcode I could use some feedback on `wasi-common::clocks` in particular.  I tried to combine that with what you wrote in `host::clocks`, but I'm not sure I captured your intent, especially in the case of `WallClock`.

BTW, I would like to keep the `WasiMonotonicClock` and `WasiSystemClock` traits rather than use `std`'s `Instant` and `SystemTime` directly, since the traits are useful for e.g. testing using simulated clocks.  We can get rid of the traits eventually once we have a mature, component-oriented way to virtualize clocks, of course.

File: wasi-common/Cargo.toml
Comment (pchickey) [inline wasi-common/Cargo.toml:13]: Let's add `publish=false` to make sure we never accidentally overwrite the "real" wasi-common

**Review by pchickey — APPROVED** [review]
This looks like a good spot to start iterating from!

**Comment by dicej** [comment]
@sunfishcode @pchickey Any concerns with merging this?

BTW, I mentioned to Pat that there was some interest in doing this development in the Wasmtime repo, but we agreed that we'd like to stick with this repo for now will we prove the concept, and tackle re-integration into the Wasmtime repo afterward.

**Comment by sunfishcode** [comment]
Looks good to me.

`wall-clock-resolution` returning `datetime` wasn't a mistake; it was to match POSIX `clock_getres` which returns a `timespec`. It's not that important though, because a `u64` of nanoseconds is enough to represent over 500 years, which is plenty precision for any practical clock.

**Comment by dicej** [comment]
> Looks good to me.
> 
> `wall-clock-resolution` returning `datetime` wasn't a mistake; it was to match POSIX `clock_getres` which returns a `timespec`. It's not that important though, because a `u64` of nanoseconds is enough to represent over 500 years, which is plenty precision for any practical clock.

Thanks for the explanation.  Yeah, it seems to me that a clock which is so imprecise that it can't measure intervals less than 500 years is probably not a clock anyone would want to use :).  OTOH, it makes sense to choose a type that can represent all possible values, so I'm happy to change it back to `datetime`.

**Comment by dicej** [comment]
I don't have merge permission on this repo, so I'll let you merge this when you see fit.  Meanwhile, since it sounds like we're on the right track, I'll continue implementing `wasi-filesystem` functions and start adding tests.

**Pull request merged**
